test/cli/pinning_remote_test.go

```diff
@@ -0,0 +1,421 @@
+			<-transitionedCh
+			res := node.IPFS("pin", "remote", "ls", "--service=svc", "--cid="+hash, "--enc=json").Stdout.String()
+			assert.Contains(t, res, hash)
+		})
+
+		t.Run("'ipfs pin remote rm --name' without --force when multiple pins match", func(t *testing.T) {
+			t.Parallel()
+			node := harness.NewT(t).NewNode().Init().StartDaemon()
+			svc, svcURL := runPinningService(t, authToken)
+			node.IPFS("pin", "remote", "service", "add", "svc", svcURL, authToken)
+
+			svc.PinAdded = func(req *pinningservice.AddPinRequest, pin *pinningservice.PinStatus) {
+				pin.M.Lock()
+				defer pin.M.Unlock()
+				pin.Status = "pinned"
+			}
+			hash := node.IPFSAddStr(string(testutils.RandomBytes(1000)))
+			node.IPFS("pin", "remote", "add", "--service=svc", "--name=force-test-name", hash)
+			node.IPFS("pin", "remote", "add", "--service=svc", "--name=force-test-name", hash)
+
+			t.Run("fails", func(t *testing.T) {
+				res := node.RunIPFS("pin", "remote", "rm", "--service=svc", "--name=force-test-name")
+				assert.Equal(t, 1, res.ExitCode())
+				assert.Contains(t, res.Stderr.String(), "Error: multiple remote pins are matching this query, add --force to confirm the bulk removal")
+			})
+
+			t.Run("matching pins are not removed", func(t *testing.T) {
+				lines := node.IPFS("pin", "remote", "ls", "--service=svc", "--name=force-test-name").Stdout.Lines()
+				assert.Contains(t, lines[0], "force-test-name")
+				assert.Contains(t, lines[1], "force-test-name")
+			})
+		})
+
+		t.Run("'ipfs pin remote rm --name --force' remove multiple pins", func(t *testing.T) {
+			t.Parallel()
+			node := harness.NewT(t).NewNode().Init().StartDaemon()
+			svc, svcURL := runPinningService(t, authToken)
+			node.IPFS("pin", "remote", "service", "add", "svc", svcURL, authToken)
+
+			svc.PinAdded = func(req *pinningservice.AddPinRequest, pin *pinningservice.PinStatus) {
+				pin.M.Lock()
+				defer pin.M.Unlock()
+				pin.Status = "pinned"
+			}
+			hash := node.IPFSAddStr(string(testutils.RandomBytes(1000)))
+			node.IPFS("pin", "remote", "add", "--service=svc", "--name=force-test-name", hash)
+			node.IPFS("pin", "remote", "add", "--service=svc", "--name=force-test-name", hash)
+
+			node.IPFS("pin", "remote", "rm", "--service=svc", "--name=force-test-name", "--force")
+			out := node.IPFS("pin", "remote", "ls", "--service=svc", "--name=force-test-name").Stdout.Trimmed()
+			assert.Empty(t, out)
+		})
+
+		t.Run("'ipfs pin remote rm --force' removes all pins", func(t *testing.T) {
+			t.Parallel()
+			node := harness.NewT(t).NewNode().Init().StartDaemon()
+			svc, svcURL := runPinningService(t, authToken)
+			node.IPFS("pin", "remote", "service", "add", "svc", svcURL, authToken)
+
+			svc.PinAdded = func(req *pinningservice.AddPinRequest, pin *pinningservice.PinStatus) {
+				pin.M.Lock()
+				defer pin.M.Unlock()
+				pin.Status = "pinned"
+			}
+			for i := 0; i < 4; i++ {
+				hash := node.IPFSAddStr(string(testutils.RandomBytes(1000)))
```

Comment:
I think you should make `IPFSAddStr`:
```go
func IPFSAdd[T ~string | ~[]byte](x T) {
  s := string(x)
  // add ...
```
This would remove a few noisy casts.

Reply:
can't have type parameters on methods sadly (or thankfully in many cases...)

Reply:
Oh I forgot about that :'(